lib/app-common.ts

```diff
@@ -53,37 +83,135 @@ function checkNodeVersion() {
 	}
 }
 
-function setupGlobalHttpProxy() {
-	// Doing this before requiring any other modules,
-	// including the 'balena-sdk', to prevent any module from reading the http proxy config
-	// before us
-	const globalTunnel = require('global-tunnel-ng');
-	const settings = require('balena-settings-client');
-	let proxy;
-	try {
-		proxy = settings.get('proxy') || null;
-	} catch (error1) {
-		proxy = null;
+export interface GlobalTunnelNgConfig {
+	host: string;
+	port: number;
+	protocol: string;
+	proxyAuth?: string;
+	connect?: string;
+	sockets?: number;
+}
+
+/**
+ * Global proxy setup. Originally, `global-tunnel-ng` was used, but it only
+ * supports Node.js versions older than 10.16.0. For v10.16.0 and later,
+ * we use `global-agent` (which only supports Node.js v10.0.0 and later).
+ *
+ * For backwards compatibility reasons, in either case we still accept a
+ * 'proxy' setting in `.balenarc.yml` that follows the
+ * `global-tunnel-ng` object configuration format:
+ *     https://www.npmjs.com/package/global-tunnel-ng#options
+ *
+ * The proxy may also be configured with the environment variables:
+ * BALENARC_PROXY, HTTP_PROXY, HTTPS_PROXY, http_proxy, and https_proxy,
+ * any of which should contain a URL in the usual format (authentication
+ * details are optional): http://username:password@domain.com:1234
+ *
+ * A proxy exclusion list in the NO_PROXY variable is only supported when
+ * `global-agent` is used, i.e. with Node.js v10.16.0 or later. The format
+ * is specified at: https://www.npmjs.com/package/global-agent#exclude-urls
+ * Patterns are matched with matcher: https://www.npmjs.com/package/matcher
+ * 'localhost' and '127.0.0.1' are always excluded. If NO_PROXY is not defined,
+ * this function adds exclusion patterns for all private IPv4 address ranges.
+ */
+async function setupGlobalHttpProxy(settings: CliSettings) {
```

Comment:
See comment above. `global-agent` fixes the Node.js support for Node 10 and Node 12 (issue #1579), **and** also provides the proxy exclusion feature requested in issue #1580.  This is why, by the way, I didn't provide separate commits for fixing the issue and adding the feature -- it's hard to separate them.